src/v/cloud_topics/level_zero/stm/ctp_stm.cc

```diff
@@ -112,8 +169,37 @@ ss::future<> ctp_stm::apply_raft_snapshot(const iobuf& buf) {
 }
 
 ss::future<iobuf> ctp_stm::take_raft_snapshot(model::offset snapshot_at) {
-    auto st = _state.get_state_at(snapshot_at);
-    co_return serde::to_iobuf(std::move(st));
+    vassert(
+      last_applied() >= snapshot_at,
+      "The snapshot is taken at offset {} but current insync offset is {}",
+      snapshot_at,
+      last_applied());
+    co_return serde::to_iobuf(_state);
+}
+
+ss::future<cluster_epoch_fence> ctp_stm::fence_epoch(cluster_epoch e) {
+    auto term = _raft->confirmed_term();
+    auto projected_epoch = _state.get_projected_epoch();
+    if (projected_epoch.has_value() && projected_epoch.value() == e) {
```

Comment:
note to self: projected == in_flight (although now i'm understanding that projected here is probably intended to mean _future_ rather than _under a function_)